# TFM_Clara_Bonacasa_Escrich
Here, you will find the scripts developped and used for my Master's Degree project in Bioinformatics and Computational Biology. For this study, several experiments were performed in order to explore hepatic cells at distint molecular levels. Most of the analyses were performed in Rstudio although Galaxy, GSEA and IPA were also used. 5 different experiments were conducted:

1- Single-cell RNA analysis of previously published data available in GEO database (GSE136103) was performed in Rstudio using Seurat package and guidelines. Here, hepatic non-parenchymal cells are studied in cirrhosis and healthy environments and we try to understand the changes in cell communities. We focus on Kupffer cells.

2- Bulk-RNA sequencing of Kupffer cells was done. Here, we have the workflow analysis to perform quality control, filtering, quantification and diffferential expression in Galaxy. The visualization part was performed in Rstudio whereas the enrichment analysis was done in IPA.

3- Single-nuclei RNA analysis of murine CD45+ cells of the liver were analyzed using as mentioned before, in Rstudio using Seurat package and guidelines. In this experiment Kupffer cells lack the expression of the scavenger receptor CD36, hence we study the impact of thhis deletion in Kupffer in lipotoxicity.

4- Proteomics analysis of in vitro Kupffer cells cultured with palmitate and LPS to mimick the steatotic environment of the liver in fatty liver disease. Here, we have the code used to visualize the results.

5- Single-nuclei RNA analysis of murine CD45+ cells of the liver were analyzed using as mentioned before, in Rstudio using Seurat package and guidelines. However, in this experiment Kupffer cells lack the expression of p38 signaling pathway by the deletion of MKK3/6. We analyze the effect it produces in Kupffer cells in steatosis.
